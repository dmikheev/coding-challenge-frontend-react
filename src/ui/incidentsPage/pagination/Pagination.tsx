import { Button } from '@material-ui/core';
import React from 'react';
import styled from 'styled-components';

const VISIBLE_PAGE_NUM_COUNT = 5;

interface IPaginationProps {
  className?: string;
  activePage: number;
  itemsCountPerPage: number;
  totalItemsCount: number;
  onChange(page: number): void;
}
const Pagination: React.FC<IPaginationProps> = ({
  className,
  activePage,
  itemsCountPerPage,
  totalItemsCount,
  onChange,
}) => {
  const numButtonsHtml: React.ReactNode[] = [];
  const visiblePageCountToOneSide = Math.floor((VISIBLE_PAGE_NUM_COUNT - 1) / 2);
  const totalPagesCount = Math.ceil(totalItemsCount / itemsCountPerPage);
  const firstPageNum = Math.max(1, activePage - visiblePageCountToOneSide);
  const lastPageNum = Math.min(totalPagesCount, activePage + visiblePageCountToOneSide);
  for (let i = firstPageNum; i <= lastPageNum; i++) {
    numButtonsHtml.push((
      <ButtonWithNum
        key={i}
        variant="contained"
        color={i === activePage ? 'secondary' : undefined}
        onClick={() => onChange(i)}
      >
        {i}
      </ButtonWithNum>
    ));
  }

  return (
    <StyledWrap className={className}>
      <ButtonWithText disabled={activePage === 1} variant="contained" onClick={() => onChange(1)}>
        {'<< First'}
      </ButtonWithText>
      <ButtonWithText
        disabled={activePage === 1}
        variant="contained"
        onClick={() => onChange(activePage - 1)}
      >
        {'< Prev'}
      </ButtonWithText>
      {numButtonsHtml}
      <ButtonWithText
        disabled={activePage === totalPagesCount}
        variant="contained"
        onClick={() => onChange(activePage + 1)}
      >
        {'Next >'}
      </ButtonWithText>
      <ButtonWithText
        disabled={activePage === totalPagesCount}
        variant="contained"
        onClick={() => onChange(totalPagesCount)}
      >
        {'Last >>'}
      </ButtonWithText>
    </StyledWrap>
  );
};
export default Pagination;

const StyledWrap = styled.div`
  & > *:not(:last-child) {
    margin-right: 20px;
  }
`;

const ButtonWithText = styled(Button)`
  min-width: 100px;
`;

const ButtonWithNum = styled(Button)`
  min-width: auto;
`;
