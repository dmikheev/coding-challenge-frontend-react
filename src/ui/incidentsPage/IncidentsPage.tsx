import { CircularProgress, Typography } from '@material-ui/core';
import React from 'react';
import styled, { css } from 'styled-components';
import { IGetIncidentsResponseResult } from '../../api/api';
import { ITEMS_PER_PAGE } from '../../constants/constants';
import * as Constants from '../../constants/constants';
import { IFilterComponentState } from './dataTypes';
import FiltersContainer from './filters/FiltersContainer';
import IncidentsList from './incidentsList/IncidentsList';
import Pagination from './pagination/Pagination';

const LOADER_SIZE = 60;

interface IIncidentsPageProps {
  className?: string;
  activePage: number | undefined;
  appliedFiltersState: IFilterComponentState;
  pageData: IGetIncidentsResponseResult | null;
  isError: boolean;
  isLoading: boolean;
  isFirstLoadDone: boolean;
  onFilterLoadData(filtersState: IFilterComponentState): void;
  onPageChange(page: number): void;
}
const IncidentsPage: React.FC<IIncidentsPageProps> = ({
  className,
  activePage,
  appliedFiltersState,
  isError,
  isLoading,
  isFirstLoadDone,
  pageData,
  onFilterLoadData,
  onPageChange,
}) => {
  let contentHtml;
  let paginationHtml;
  if (isError) {
    contentHtml = <ErrorContent/>;
  } else {
    contentHtml = <IncidentsList data={pageData}/>;
    paginationHtml = (pageData && pageData.total > Constants.ITEMS_PER_PAGE) && (
      <StyledPagination
        totalItemsCount={pageData.total}
        activePage={activePage || 1}
        itemsCountPerPage={ITEMS_PER_PAGE}
        onChange={onPageChange}
      />
    );
  }

  const loaderHtml = isLoading && (
    <LoaderDimmerWrap withBackground={isFirstLoadDone}>
      <StyledLoader size={LOADER_SIZE}/>
    </LoaderDimmerWrap>
  );

  return (
    <div className={className}>
      <StyledFiltersContainer
        appliedFiltersState={appliedFiltersState}
        onLoadData={onFilterLoadData}
      />
      <ContentWrap>
        {contentHtml}
        {loaderHtml}
      </ContentWrap>
      {paginationHtml}
    </div>
  );
};
export default IncidentsPage;

const StyledFiltersContainer = styled(FiltersContainer)`
  margin-bottom: 20px;
`;

const ContentWrap = styled.div`
  position: relative;
  min-height: 200px;
`;

const StyledPagination = styled(Pagination)`
  margin-top: 30px;
`;

interface ILoaderWrapProps {
  withBackground: boolean;
}
const LoaderDimmerWrap = styled.div<ILoaderWrapProps>`
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
    
  ${props => props.withBackground && css`
    background-color: rgba(0,0,0,0.4);
    box-shadow: 
      0 1px 5px 0 rgba(0,0,0,0.2), 
      0 2px 2px 0 rgba(0,0,0,0.14), 
      0 3px 1px -2px rgba(0,0,0,0.12);
  `}
`;
const StyledLoader = styled(CircularProgress)`
  position: sticky;
  top: calc(50% - ${Math.round(LOADER_SIZE / 2)}px);
  display: block;
  margin: 74px auto 78px;
`;

const ErrorContent: React.FC = () => (
  <Typography color="error">Ooops, something went wrong</Typography>
);
