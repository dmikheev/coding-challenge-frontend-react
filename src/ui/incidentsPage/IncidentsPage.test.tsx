import MomentUtils from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import moment from 'moment';
import React from 'react';
import IncidentsPage, { IIncidentsPageProps } from './IncidentsPage';
import Pagination from './pagination/Pagination';

function nop() {}
const testProps: IIncidentsPageProps = {
  activePage: 2,
  appliedFiltersState: {
    query: 'test',
    dateFrom: moment.utc('2019-01-01'),
    dateTo: moment.utc('2019-02-01'),
  },
  pageData: {
    incidents: [
      { id: 1, title: 'Test title 1',
        description: 'Detailed test description 1',
        address: 'Test address 1',
        occurred_at: Date.UTC(2019, 0, 1),
        updated_at: Date.UTC(2019, 1, 1),
        source: {
          html_url: 'https://test.com/bike/1',
        },
        media: {
          image_url_thumb: 'https://test.com/bike/1/photo.jpg',
        },
      },
    ],
    total: 11,
  },
  isError: false,
  isLoading: false,
  isFirstLoadDone: true,
  onFilterLoadData: nop,
  onPageChange: nop,
};

it('renders correctly with results', () => {
  const wrapper = shallow(
    <IncidentsPage {...testProps}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly during the first loading', () => {
  const filtersState = {
    query: '',
    dateFrom: null,
    dateTo: null,
  };
  const wrapper = shallow(
    <IncidentsPage
      activePage={undefined}
      appliedFiltersState={filtersState}
      pageData={null}
      isError={false}
      isLoading={true}
      isFirstLoadDone={false}
      onFilterLoadData={nop}
      onPageChange={nop}
    />
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly during next loadings', () => {
  const wrapper = shallow(
    <IncidentsPage {...testProps} isLoading={true}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly if active page is not defined', () => {
  const wrapper = mount(
    <MuiPickersUtilsProvider utils={MomentUtils}>
      <IncidentsPage {...testProps} activePage={undefined}/>
    </MuiPickersUtilsProvider>
  );

  const pagination = wrapper.find(Pagination);
  expect(pagination).toHaveLength(1);
  expect(pagination.props().activePage).toBe(1);
});

it('shows error message on error', () => {
  const wrapper = shallow(
    <IncidentsPage {...testProps} isError={true}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('shows "no results" message if incidents data is empty', () => {
  const pageData = { incidents: [], total: 0 };
  const wrapper = shallow(
    <IncidentsPage {...testProps} pageData={pageData}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
