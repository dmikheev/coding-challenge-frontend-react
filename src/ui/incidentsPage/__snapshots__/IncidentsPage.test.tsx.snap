// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly during next loadings 1`] = `
<div>
  <Styled(FiltersContainer)
    appliedFiltersState={
      Object {
        "dateFrom": "2019-01-01T00:00:00.000Z",
        "dateTo": "2019-02-01T00:00:00.000Z",
        "query": "test",
      }
    }
    onLoadData={[Function]}
  />
  <styled.div>
    <IncidentsList
      data={
        Object {
          "incidents": Array [
            Object {
              "address": "Test address 1",
              "description": "Detailed test description 1",
              "id": 1,
              "media": Object {
                "image_url_thumb": "https://test.com/bike/1/photo.jpg",
              },
              "occurred_at": 1546300800000,
              "source": Object {
                "html_url": "https://test.com/bike/1",
              },
              "title": "Test title 1",
              "updated_at": 1548979200000,
            },
          ],
          "total": 11,
        }
      }
    />
    <styled.div
      withBackground={true}
    >
      <Styled(WithStyles(ForwardRef(CircularProgress)))
        size={60}
      />
    </styled.div>
  </styled.div>
  <Styled(Pagination)
    activePage={2}
    itemsCountPerPage={10}
    onChange={[Function]}
    totalItemsCount={11}
  />
</div>
`;

exports[`renders correctly during the first loading 1`] = `
<div>
  <Styled(FiltersContainer)
    appliedFiltersState={
      Object {
        "dateFrom": null,
        "dateTo": null,
        "query": "",
      }
    }
    onLoadData={[Function]}
  />
  <styled.div>
    <styled.div
      withBackground={false}
    >
      <Styled(WithStyles(ForwardRef(CircularProgress)))
        size={60}
      />
    </styled.div>
  </styled.div>
</div>
`;

exports[`renders correctly with results 1`] = `
<div>
  <Styled(FiltersContainer)
    appliedFiltersState={
      Object {
        "dateFrom": "2019-01-01T00:00:00.000Z",
        "dateTo": "2019-02-01T00:00:00.000Z",
        "query": "test",
      }
    }
    onLoadData={[Function]}
  />
  <styled.div>
    <IncidentsList
      data={
        Object {
          "incidents": Array [
            Object {
              "address": "Test address 1",
              "description": "Detailed test description 1",
              "id": 1,
              "media": Object {
                "image_url_thumb": "https://test.com/bike/1/photo.jpg",
              },
              "occurred_at": 1546300800000,
              "source": Object {
                "html_url": "https://test.com/bike/1",
              },
              "title": "Test title 1",
              "updated_at": 1548979200000,
            },
          ],
          "total": 11,
        }
      }
    />
  </styled.div>
  <Styled(Pagination)
    activePage={2}
    itemsCountPerPage={10}
    onChange={[Function]}
    totalItemsCount={11}
  />
</div>
`;

exports[`shows "no results" message if incidents data is empty 1`] = `
<div>
  <Styled(FiltersContainer)
    appliedFiltersState={
      Object {
        "dateFrom": "2019-01-01T00:00:00.000Z",
        "dateTo": "2019-02-01T00:00:00.000Z",
        "query": "test",
      }
    }
    onLoadData={[Function]}
  />
  <styled.div>
    <Styled(WithStyles(ForwardRef(Typography)))>
      No results
    </Styled(WithStyles(ForwardRef(Typography)))>
  </styled.div>
</div>
`;

exports[`shows error message on error 1`] = `
<div>
  <Styled(FiltersContainer)
    appliedFiltersState={
      Object {
        "dateFrom": "2019-01-01T00:00:00.000Z",
        "dateTo": "2019-02-01T00:00:00.000Z",
        "query": "test",
      }
    }
    onLoadData={[Function]}
  />
  <styled.div>
    <Styled(WithStyles(ForwardRef(Typography)))
      color="error"
    >
      Ooops, something went wrong
    </Styled(WithStyles(ForwardRef(Typography)))>
  </styled.div>
</div>
`;
